driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import com.mongodb.assertions.Assertions;
+import org.bson.BsonArray;
+import org.bson.BsonBoolean;
+import org.bson.BsonDateTime;
+import org.bson.BsonDocument;
+import org.bson.BsonDouble;
+import org.bson.BsonInt32;
+import org.bson.BsonInt64;
+import org.bson.BsonNull;
+import org.bson.BsonString;
+import org.bson.BsonValue;
+import org.bson.conversions.Bson;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+
+/**
+ * Convenience methods related to {@link Expression}.
+ */
+public final class Expressions {
+
+    private Expressions() {}
+
+    /**
+     * Returns an expression having the same boolean value as the provided
+     * boolean primitive.
+     *
+     * @param of the boolean primitive
+     * @return the boolean expression
+     */
+    public static BooleanExpression of(final boolean of) {
+        // we intentionally disallow ofBoolean(null)
+        return new MqlExpression<>((codecRegistry) -> new BsonBoolean(of));
+    }
+
+    /**
+     * Returns an expression having the same integer value as the provided
+     * int primitive.
+     *
+     * @param of the int primitive
+     * @return the integer expression
+     */
+    public static IntegerExpression of(final int of) {
+        return new MqlExpression<>((codecRegistry) -> new BsonInt32(of));
+    }
+    public static IntegerExpression of(final long of) {
+        return new MqlExpression<>((codecRegistry) -> new BsonInt64(of));
+    }
+    public static NumberExpression of(final double of) {
+        return new MqlExpression<>((codecRegistry) -> new BsonDouble(of));
+    }
+    public static DateExpression of(final Instant of) {
+        return new MqlExpression<>((codecRegistry) -> new BsonDateTime(of.toEpochMilli()));
+    }
+
+    /**
+     * Returns an expression having the same string value as the provided
+     * string.
+     *
+     * @param of the string
+     * @return the string expression
+     */
+    public static StringExpression of(final String of) {
+        Assertions.notNull("String", of);
+        return new MqlExpression<>((codecRegistry) -> new BsonString(of));
+    }
```

Comment:
The use of `Assertions.notNull` to check arguments in public API methods is inconsistent. The check should either be done in all API methods in the PR, or in none.

Reply:
It seems that we could skip methods which would cause an NPE regardless?

Reply:
This suggestion is conceptually the same as the following more general approach: _"if an invalid method argument always causes a runtime error (given the current state of the code), there is no need to validate the argument"_, which is a bad idea.

Just in case, the proposed approach (they are generalizable and are equally applicable to the more general approach described above):

1. Explicit checks provide an error message that explicitly points out to a user mistake.
2. Without an explicit check the NPE may, generally speaking, happen somewhere deep in the call stack that originates from method where an explicit check should have been put.
3. Placing the checks selectively burdens readers with the task of repeating the manual code analysis to figure out if the code fails with an NPE despite not having an explicit check, or there is a bug in the code.
4. Even if one manages to analyze all the current code correctly (which is questionable), future changes may, generally speaking, result in the method stopping to produce an NPE and starting to behave incorrectly in some other way.

Reply:
Fair enough. I need to think of how to reliably test this.